checker/src/org/checkerframework/checker/units/UnitsAnnotatedTypeFactory.java

```diff
@@ -326,17 +467,376 @@ public Void visitCompoundAssignment(CompoundAssignmentTree node, AnnotatedTypeMi
      * A class for adding annotations based on tree
      */
     private class UnitsTreeAnnotator extends TreeAnnotator {
+        // cache expressions evaluated in evalMathExpression
+        private final Map<String, Pair<Double, Boolean>> mathExpressions;
+
+        private final TypeMirror stringType;
+        private final TypeMirror mathType;
+        private final TypeMirror strictMathType;
 
         UnitsTreeAnnotator(UnitsAnnotatedTypeFactory atypeFactory) {
             super(atypeFactory);
+
+            mathExpressions = new HashMap<String, Pair<Double, Boolean>>();
+
+            stringType = getElementUtils().getTypeElement(java.lang.String.class.getCanonicalName()).asType();
+            mathType = getElementUtils().getTypeElement(java.lang.Math.class.getCanonicalName()).asType();
+            strictMathType = getElementUtils().getTypeElement(java.lang.StrictMath.class.getCanonicalName()).asType();
+        }
+
+        private boolean isSameUnderlyingType(TypeMirror lht, TypeMirror rht) {
+            // use typeUtils.isSameType instead of TypeMirror.equals as this
+            // will check only the underlying type and ignores declarations on
+            // the type mirror
+            return checker.getTypeUtils().isSameType(lht, rht);
         }
 
         @Override
-        public Void visitBinary(BinaryTree node, AnnotatedTypeMirror type) {
-            AnnotatedTypeMirror lht = getAnnotatedType(node.getLeftOperand());
-            AnnotatedTypeMirror rht = getAnnotatedType(node.getRightOperand());
-            Tree.Kind kind = node.getKind();
+        public Void visitMethodInvocation(MethodInvocationTree node, AnnotatedTypeMirror type) {
+            Name methodName = TreeUtils.methodName(node);
+            AnnotatedTypeMirror receiverType = getReceiverType(node);
+
+            // If there's no receivers, then return immediately
+            if (receiverType == null) {
+                return null;
+            }
+
+            // check to see if the class is java.lang.Math or
+            // java.lang.StrictMath
+            TypeMirror underlyingType = receiverType.getUnderlyingType();
+
+            // java.lang.StrictMath implements the same methods as
+            // java.lang.Math except for incrementExact, decrementExact, and
+            // negateExact
+            // The common methods have the same signature and same expected
+            // units
+            boolean isMathType = isSameUnderlyingType(underlyingType, mathType) || isSameUnderlyingType(underlyingType, strictMathType);
+            // see which method in the Math library was called
+            boolean isAddExact = methodName.contentEquals("addExact");
```

Comment:
Why compute all these booleans, for any method call, even if `isMathType` might be false and no further processing fill happen? Can't you just move each boolean to the single use in the corresponding condition and safe all this work?